### PR TITLE
Automatically re-subscribe to all existing channels when reconnected

### PR DESCRIPTION
**Pull request by RobertoSchneiders**

### CODE REVIEW

File: PusherClient/EventEmitter.cs [hunk 1]
```diff
@@ -16,7 +16,9 @@ public void Bind(string eventName, Action<dynamic> listener)
         {
             if(_eventListeners.ContainsKey(eventName))
             {
-                _eventListeners[eventName].Add(listener);
+                // To avoid triggering the same bind multiple times
```
Comment (leggetter) [inline PusherClient/EventEmitter.cs]: This comment helps explain what this does. But I'm not sure of the why. Why was this required? Out of interest, more than anything else.

File: PusherClient/Pusher.cs [hunk 2]
```diff
@@ -130,7 +130,7 @@ public Channel Subscribe(string channelName)
             if (_connection.State != ConnectionState.Connected)
                 throw new PusherException("You must wait for Pusher to connect before you can subscribe to a channel", ErrorCodes.NotConnected);
```
Comment (leggetter) [inline PusherClient/Pusher.cs]: With the change [here](https://github.com/pusher-community/pusher-websocket-dotnet/pull/9/files#diff-25180fb73bd9f8ee4a11d4b830a8bec4R261) it would seem that the `PusherException` is no longer required.

**Comment by leggetter** [comment]
The logic seems pretty sound. Maybe the PR could be updated as pointed out in https://github.com/pusher-community/pusher-websocket-dotnet/pull/9#discussion-diff-40000778

Then I'll look at manually testing before merging into master and pushing a new version to Nuget.

File: PusherClient/Pusher.cs (hunk 2 above)
Comment (RobertoSchneiders) [inline PusherClient/Pusher.cs]: I'm not sure. A code like this will break with this exception:
```
pusher = new Pusher("key");
pusher.Connect();
pusher.Subscribe("chat");
```

Comment (leggetter) [inline PusherClient/Pusher.cs]: There probably are reasons that making this change would break things. Maybe around a Channel actually being in a `Subscribing` state - something we don't have right now.

So, let's ignore my comment and I'll manually test as-is and merge etc.

Thanks :+1:

File: PusherClient/EventEmitter.cs (hunk 1 above)
Comment (RobertoSchneiders) [inline PusherClient/EventEmitter.cs]: In my case (as also on ExampleApplication), I have a code like this on Subscribed event.

```
static void _chatChannel_Subscribed(object sender)
{
    Console.WriteLine("Hi! Type 'quit' to exit, otherwise type anything to chat!");

    chatChannel.Bind("my-message", (dynamic data) =>
    {
        Console.WriteLine("[" + data.name + "] " + data.message);
    });
}
``` 

When some connection failure occurs, all channels are subscribed again. Which means that the Subscribed event can be triggered infinite times. Every time this happens a new bind is created for the same method. So, when a message is received it triggers the event/method multlipe times. 

In this particulare example, it writes the same message multiple times on the console 

Should I remove this comment? Explain this in details?

Comment (RobertoSchneiders) [inline PusherClient/EventEmitter.cs]: I think it will be better have only one bind per eventname, but I didn't implemented it because it can break some existing applications.

Comment (leggetter) [inline PusherClient/EventEmitter.cs]: I wouldn't expect you to need to bind to events on a channel after subscription. You should be able to:

```cs
var pusher = new Pusher(...);
pusher.Connect();

// ... wait for connection (although you shouldn't really need to have to do that - but we'll keep that for another day)

var chatChannel = pusher.Subscribe("chat");
chatChannel.Bind("my-message", (dynamic data) =>
    {
        Console.WriteLine("[" + data.name + "] " + data.message);
    });
```

So, to me it seems the [code in the ExampleApplication](https://github.com/pusher-community/pusher-websocket-dotnet/blob/master/ExampleApplication/Program.cs#L101-L104) is causing confusion and is doing something a bit strange.

**hands up** - it has been a while since I've looked at this library. Hopefully I'll have a chance to get back up to speed as part of looking at this pull request.

Comment (leggetter) [inline PusherClient/EventEmitter.cs]: We need to support more than one bind per event name because multiple pieces of application logic may want to know about the same event e.g. things like connection events, which also uses the EventEmitter.

Comment (RobertoSchneiders) [inline PusherClient/EventEmitter.cs]: About the EventEmitter, I got it. I am not familiar with this approach, but, it makes sense.

About the bindings and your example, how much you will wait for the connection? 
It is undetermined, right? 

I think thats the reason why the example use the events to subscribe to the channel and bind the methods. 

How you will do the "wait for the connection"? A loop testing the status?

**Comment by leggetter** [comment]
68d141335f102e515d2e0f63d36a4104c617fc58 pull in the majority of the work here

File: PusherClient/EventEmitter.cs (hunk 1 above)
Comment (leggetter) [inline PusherClient/EventEmitter.cs]: You don't need to "wait" for the connection. You have a connection "event" handler which informs you when the connection state changes.
